.Lib9c.Tests/Action/DPoS/Control/SlashCtrlComplexTest.cs

```diff
@@ -0,0 +1,588 @@
+namespace Lib9c.Tests.Action.DPoS.Control
+{
+    using System;
+    using System.Linq;
+    using Libplanet.Action.State;
+    using Libplanet.Crypto;
+    using Nekoyume.Action.DPoS.Control;
+    using Nekoyume.Action.DPoS.Misc;
+    using Nekoyume.Action.DPoS.Model;
+    using Xunit;
+
+    public class SlashCtrlComplexTest : SlashCtrlTestBase
+    {
+        public static readonly object[][] TestData
+            = Enumerable.Range(0, Random.Shared.Next(5, 10))
+                        .Select(item => new object[]
+                        {
+                            Random.Shared.Next() % 2 == 0, // jailed
+                            Random.Shared.Next(1, 101), // slashFactor
+                        })
+                        .ToArray();
+
+        private const int ValidatorCount = 2;
+        private static readonly int DelegatorCount = Random.Shared.Next(2, 32);
+
+        private readonly PublicKey[] _operatorPublicKeys;
+        private readonly Address[] _operatorAddresses;
+        private readonly Address[] _delegatorAddresses;
+        private readonly Address[] _validatorAddresses;
+
+        private IWorld _states;
+
+        public SlashCtrlComplexTest()
+        {
+            _operatorPublicKeys = Enumerable.Range(0, ValidatorCount)
+                .Select(_ => new PrivateKey().PublicKey)
+                .ToArray();
+            _operatorAddresses = _operatorPublicKeys.Select(item => item.Address).ToArray();
+            _delegatorAddresses = Enumerable.Range(0, DelegatorCount)
+                .Select(_ => CreateAddress())
+                .ToArray();
+            _validatorAddresses = _operatorAddresses
+                .Select(item => Validator.DeriveAddress(item))
+                .ToArray();
+            _states = InitializeStates();
+        }
+
+        [Theory]
+        [MemberData(nameof(TestData))]
+        public void Slash_Test(bool jailed, int slashFactor)
+        {
+            // Given
+            var validatorNCG = NextNCG();
+            var validatorPower = PowerFromNCG(validatorNCG);
+            var validatorShare = ShareFromPower(validatorPower);
```

Comment:
IMO, it's better to obtain `validatorPower` and `validatorShare` from states via `GetPower()` or `GetShare()` right after validator promotion, considering future modification on tests.
Actually, there are lots of lines like this, which seems to be fragile for me, considering test extension(for example, on the condition that inflation is added right after promotion) but only for this small unit test context, okay.

Reply:
Except for `validatorNCG`, `validatorPower` and `validatorShare` are defined as expected values to be used in tests. Therefore, if a test passes with the expected and actual values being equal, I believe there is no fragility.